Document exporting certificates from PKCS#12 keystores (#1403)
elastic/rally: docs/command_line_reference.rst

**TLS/SSL Certificate Verification**

Server certificate verification is controlled with the ``verify_certs`` boolean. The default value is `true`. To disable use ``verify_certs:false``.
If ``verify_certs:true``, Rally will attempt to verify the certificate provided by Elasticsearch. If they are private certificates, you will also need to supply the private CA certificate in PEM format using ``ca_certs:'/path/to/cacert.pem'``.

You can also optionally present client certificates, e.g. if Elasticsearch has been configured with ``xpack.security.http.ssl.client_authentication: required`` (see also `Elasticsearch HTTP TLS/SSL settings <https://www.elastic.co/guide/en/elasticsearch/reference/current/security-settings.html#http-tls-ssl-settings>`_).
Client certificates can be presented regardless of the ``verify_certs`` setting, but it's strongly recommended to always verify the server certificates.

> **CA & Client Certificate Format**

Reviewer: I still have to test this, but can you please move the instructions after then "TLS/SSL Examples" section? Otherwise it isn't clear why we're talking about client_key and client_cert.

Author: > can you please move the instructions after then "TLS/SSL Examples" section?

👍🏼 Will do. I had actually considered the examples section initially, then thought it might be better to have client_cert and client_key before the examples.

Reviewer: Ah, I see what you mean, but for better or worse, those examples are the only part where we currently define them. (Another option is to define them properly in the "Certificate Verification" section.)

That way, we can keep separate the **definition** of the option from their **conversion** from PKCS#12.

Author: Adjustments are in https://github.com/elastic/rally/pull/1403/commits/ee28cd159f7e5f5a42affea6c80cf251be51f5bd.

Rally supports PEM format for CA and client certificates. Certificates in PKCS#12 formatted keystores will need to be exported to PEM format in order for Rally to use them. The ``openssl`` command can be used to export certificates from a PKCS#12 keystore. In the following example, PKCS#12 file ``elastic-certificates.p12`` was generated by the Elasticsearch ``elasticsearch-certutil`` command.

* Export CA certificate:

 ::

>    openssl pkcs12 -in elastic-certificates.p12 -cacerts -out cacert.pem

Reviewer: Let's use the default `elasticsearch-certutil` file name?
```suggestion
   openssl pkcs12 -in elastic-stack-ca.p12 -cacerts -out cacert.pem
```

Author: `elastic-certificates.p12` is the default name of the single-instance bundle generated by `elasticsearch-certutil` and it always includes the CA certificate and not the CA private key. It was probably the wrong choice. The `-cacerts` option only applies to instance bundles generated by certutil, so pulling it from the CA bundle is a little simpler:

`openssl pkcs12 -in elastic-stack-ca.p12 -out cacert.pem`

I reason the CA bundle file name would be the most likely to remain the same and a user knowing their way around `openssl` can export the CA cert however they want. I will change it.

Reviewer: I'm not sure I'm following. Where are you seeing `elastic-certificates.p12`? With the 7.16 and 8.0.0-beta1 Docker images, `elasticsearch-certutil ca` generates a `elastic-stack-ca.p12` file.

Also, have you tested that Rally can use those PEM files correctly?

Author: > I'm not sure I'm following. Where are you seeing `elastic-certificates.p12`? With the 7.16 and 8.0.0-beta1 Docker images, `elasticsearch-certutil ca` generates a `elastic-stack-ca.p12` file.

`elastic-certificates.p12` is the default file produced when executing the `cert` subcommand for `elasticsearch-certutil`. See the example at https://www.elastic.co/guide/en/elasticsearch/reference/current/certutil.html#_examples_5.

It was worded this way to encourage the use of a node certificate vs. the CA certificate for clients. And good point, I will definitely verify the exported certificates work with Rally correctly.

Please let me know if anything is unclear.

Author: I can confirm certificates exported using the `openssl` commands in this PR work correctly with Rally without further modification. The only consideration probably deserving further documentation is `verify_certs:true` since the hostname used in `--target-hosts` must match the one assigned to the node's instance certificate to complete server verification. In the default case for a single node, `elasticsearch-certutil` assigns the instance name "instance". 

```
openssl x509 -in client_cert.pem -noout -subject
subject= /CN=instance
```

All of this can get much more complex and I want to keep these docs as simple as possible to serve as a starting point for those looking to use server certificate verification and client authentication. That said, a better option and practice for a Rally cert/key would be to generate one specifically for benchmarking:

```sh
bin/elasticsearch-certutil cert --ca elastic-stack-ca.p12 --dns rally --pem --name rally
```

I look forward to updating the PR pending your feedback @pquentin.

Reviewer: Sorry for the delay in reviewing, and thanks for iterating!

I agree that the idea is just to give instructions for the simplest use case, we don't have to document more complex cases. If documenting `--name rally` helps, then we can do it, but otherwise leave it like this.

While this now looks good, I have not managed to test this, because I can't make Elasticsearch work with p12 files. https://www.elastic.co/guide/en/elasticsearch/reference/current/configuring-tls-docker.html is actually using PEM (turns out there's a `--pem` flag!) and https://www.elastic.co/guide/en/elasticsearch/reference/7.16/security-basic-setup-https.html suggests using `elasticsearch-certutil http` instead, and requires more fiddling than I'd like.

What steps have you followed?

**TLS/SSL Certificate Verification**

Server certificate verification is controlled with the ``verify_certs`` boolean. The default value is `true`. To disable use ``verify_certs:false``.

> If ``verify_certs:true``, Rally will attempt to verify the certificate provided by Elasticsearch. If they are private self-signed certificates, you will also need to supply the private self-signed CA certificate in PEM format using ``ca_certs:'/path/to/cacert.pem'``.

Reviewer: The wording was incorrect, even before this PR. The server (Elastic) does not present a self-signed a certificate, it presents a server certificate that was signed by a root CA. And yes, all root CA are self-signed, but I think "private" is a better description than "self-signed".

```suggestion
If ``verify_certs:true``, Rally will attempt to verify the certificate provided by Elasticsearch. If this certificate is signed by a private certificate authority (CA), you will also need to supply that CA in PEM format using ``ca_certs:'/path/to/cacert.pem'``.
```

You don't have to agree, please merge with or without this change. It's fine to merge a doc change on Friday.

Author: I like the revised wording, thank you!

* Enable SSL, verify server certificates using private CA: ``--client-options="use_ssl:true,verify_certs:true,ca_certs:'/path/to/cacert.pem'"``
* Enable SSL, verify server certificates using private CA, present client certificates: ``--client-options="use_ssl:true,verify_certs:true,ca_certs:'/path/to/cacert.pem',client_cert:'/path/to/client_cert.pem',client_key:'/path/to/client_key.pem'"``

**Converting PKCS#12 Files to PEM Format**

Rally supports PEM format for CA and client certificates. Certificates in PKCS#12 formatted keystores will need to be exported to PEM format in order for Rally to use them. The ``openssl`` command can be used to export certificates from a PKCS#12 keystore. In the following example, PKCS#12 file ``elastic-stack-ca.p12`` is generated by the Elasticsearch ``elasticsearch-certutil ca`` command and contains only the CA certificate and private key.

* Export CA certificate:

 ::

>    openssl pkcs12 -in elastic-stack-ca.p12 -out cacert.pem

Author: I missed it since openssl on macOS behaves slightly differently than Linux when an export password is not supplied. Sorry for that, I've added `-nokeys` to correct it.
